### PR TITLE
Openflow_acl_test.xxx.ftp: Create a New FTP Site by Editing the IIS Configuration Files

### PR DESCRIPTION
**Pull request by quanwenli**
id: 1475696

Pkgmgr is deprecated, instead use dism. and Create a New FTP Site
by Editing the IIS Configuration Files

Signed-off-by: Wenli Quan <wquan@redhat.com>

### CODE REVIEW

**Comment by quanwenli** [comment]
JOB ID     : e97335b6ce045f4014678a5383a07537176dbdc0
JOB LOG    : /root/avocado/job-results/job-2018-06-05T04.51-e97335b/job.log
 (1/1) Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.openflow_acl_test.access_deny_special_target.: PASS (318.27 s)

JOB ID     : d02c02ee3a667ddb8b52307be64465b51cf80a71
JOB LOG    : /root/avocado/job-results/job-2018-06-07T03.12-d02c02e/job.log
 (1/1) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win7.x86_64.sp1.io-github-autotest-qemu.openflow_acl_test.access_deny_special_target.ftp: PASS (293.65 s)

JOB ID     : 9d201bb4bc34efa26910e44d0138b0d83e558d87
JOB LOG    : /root/avocado/job-results/job-2018-06-07T03.22-9d201bb/job.log
 (1/1) Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win8.i386.1.io-github-autotest-qemu.openflow_acl_test.access_deny_special_target.ftp: PASS (211.88 s)

**Comment by quanwenli** [comment]
@xiywang could you help review it ? thx

**Comment by quanwenli** [comment]
@vivianQizhu, could you review this? Thanks

**Comment by vivianQizhu** [comment]
Code looks good, @xiywang Would you help provide latest test result? Thanks.

**Comment by xiywang** [comment]
# git am 1415.patch
Applying: Openflow_acl_test.xxx.ftp: Create a New FTP Site by Editing the IIS Configuration Files
/home/verify/tp-qemu/.git/rebase-apply/patch:21:  trailing whitespace.

/home/verify/tp-qemu/.git/rebase-apply/patch:33:  trailing whitespace.

/home/verify/tp-qemu/.git/rebase-apply/patch:36:  trailing whitespace.

/home/verify/tp-qemu/.git/rebase-apply/patch:39:  trailing whitespace.

/home/verify/tp-qemu/.git/rebase-apply/patch:46:  trailing whitespace.

warning: 5 lines add whitespace errors.

@quanwenli The patch cannot be applied successfully, please check and modify it again. Thanks.

**Comment by quanwenli** [comment]
@xiywang I fix the errors, it's because FTP server configuration file does have have indent requirement.

**Comment by quanwenli** [comment]
@vivianQizhu @xiywang I just add a commit for support python3. please help review. and test are all pass with openvwitch avocado-vt patch #1579.

JOB LOG    : /root/avocado/job-results/job-2018-11-05T00.55-6dc5750/job.log
 (1/4) Host_RHEL.m8.u0.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.openflow_acl_test.access_deny_all_target.ftp: PASS (157.57 s)
 (2/4) Host_RHEL.m8.u0.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.openflow_acl_test.access_deny_special_target.ftp: PASS (332.58 s)
 (3/4) Host_RHEL.m8.u0.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.openflow_acl_test.accessed_deny_all_target.ftp: PASS (344.82 s)
 (4/4) Host_RHEL.m8.u0.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.openflow_acl_test.accessed_deny_special_target.ftp: PASS (315.93 s)

@xiywang and for DHCP issue, it's better to add " --netdst=ovs0 " in loop

**Comment by xiywang** [comment]
The code looks good. I'll run the case again and provide the log later.

**Comment by vivianQizhu** [comment]
Depends on https://github.com/avocado-framework/avocado-vt/pull/1579

**Comment by xiywang** [comment]
(1/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.my.openflow_acl_test.access_deny_all_target.ftp: PASS (156.15 s)
 (2/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.my.openflow_acl_test.access_deny_special_target.ftp: ERROR: Login timeout expired    (output: "exceeded 360 s timeout, last failure: Client process terminated    (status: 1,    output: 'Ncat: No route to host.\\n')") (391.16 s)
 (3/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.my.openflow_acl_test.accessed_deny_all_target.ftp: ERROR: Login timeout expired    (output: "exceeded 360 s timeout, last failure: Client process terminated    (status: 1,    output: 'Ncat: No route to host.\\n')") (392.08 s)
 (4/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.my.openflow_acl_test.accessed_deny_special_target.ftp: ERROR: Login timeout expired    (output: "exceeded 360 s timeout, last failure: Client process terminated    (status: 1,    output: 'Ncat: No route to host.\\n')") (391.66 s)

@quanwenli Could you help check again?

**Comment by quanwenli** [comment]
@xiywang, you should add like " --netdst=ovs0" in staf.

**Comment by quanwenli** [comment]
@xiywang sorry, should be --netdst=atbr0 in staf. 2-4 should be run with private atbr0, 1 could be run atbr0 or ovs. but it may some problem when in one loop have atbr0/ovs0 together.

**Comment by xiywang** [comment]
@quanwenli If so, do you think we should fix this issue? Because it's not reasonable to run the same loop both with --netdst=ovs0 and --netdst=atbr0. When some cases failed, then we try with another netdst.
It's not related to this bug, but we need to consider this problem.
I'll run with netdst=atbr0 and provide the log later.

**Comment by xiywang** [comment]
(1/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.my.openflow_acl_test.access_deny_all_target.ftp: ERROR: No ipv4 DHCP lease for MAC 9a:dd:de:df:e0:e1 (400.06 s)
 (2/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.my.openflow_acl_test.access_deny_special_target.ftp: ERROR: No ipv4 DHCP lease for MAC 9a:dd:de:df:e0:e1 (407.97 s)
 (3/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.my.openflow_acl_test.accessed_deny_all_target.ftp: ERROR: No ipv4 DHCP lease for MAC 9a:dd:de:df:e0:e1 (408.26 s)
 (4/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.my.openflow_acl_test.accessed_deny_special_target.ftp: ERROR: No ipv4 DHCP lease for MAC 9a:dd:de:df:e0:e1 (408.41 s)



Here's the log while using --netdst=atbr0. @quanwenli

**Comment by quanwenli** [comment]
＠xiywang, could you check debug.log to see if there is a ip on atbr0 and dhcp server(dnsmasq) is running on atbr0's ip. could you run job again and meanwhile to checking atbr0, dhcp to see why guest could not get ip. 
like this guest could not get ip , it's really not this patch issue.

**Comment by quanwenli** [comment]
> @quanwenli If so, do you think we should fix this issue? Because it's not reasonable to run the same loop both with --netdst=ovs0 and --netdst=atbr0. When some cases failed, then we try with another netdst.
> It's not related to this bug, but we need to consider this problem.
> I'll run with netdst=atbr0 and provide the log later.

one of the possible is that the physical card was bind to ovs0 firstly in case 1, then in case 2-4, the physical card was bind to atbr0 again but without unbind from ovs0. so if you would like to support ovs0/atbr0 in one loop, it maybe need two physical cards totally, one is bind to ovs0 , the other one is bind to atbr0.

**Comment by quanwenli** [comment]
@xiywang @vivianQizhu do you have any comment, could we merge this request.

**Comment by xiywang** [comment]
(1/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.my.openflow_acl_test.access_deny_all_target.ftp: CANCEL: Can not access localhost at the beginning. Please check your setup.Command: echo get autotest-test > C:\ftp.txt && echo quit >> C:\ftp.txt && ftp -n -s:C:\ftp.txt 10.73.72.146. Output: > ftp: connect :Connection timed out\nftp> get autotest-test  \nNot conn... (162.49 s)
 (2/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.my.openflow_acl_test.access_deny_special_target.ftp: ERROR: [Errno 2] No such file or directory: '/var/tmp/avocado_zu7a445w/dnsmasq.pid' (1.42 s)
 (3/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.my.openflow_acl_test.accessed_deny_all_target.ftp: ERROR: Login timeout expired    (output: "exceeded 360 s timeout, last failure: Client process terminated    (status: 1,    output: 'Ncat: No route to host.\\n')") (405.77 s)
 (4/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.my.openflow_acl_test.accessed_deny_special_target.ftp: ERROR: Login timeout expired    (output: "exceeded 360 s timeout, last failure: Client process terminated    (status: 1,    output: 'Ncat: No route to host.\\n')") (411.03 s)


The last two errors should be caused by an known bug of snapshot=on the two vms will get the same ip. But the second case is failed because of dnsmasq service is not running but in avocado-vt, the code tried to stop the service. And the first one looks like the ftp server is not connectable.

@quanwenli Could you check the first two cases again? Thanks.

**Comment by quanwenli** [comment]
@xiywang In case access_deny_special_target,accessed_deny_all_target,accessed_deny_special_target, they all have "netdst = private", so it's better add "--netdst=atbr0" in your loop. 

I try again, the four tests are all passed. I update the results to bug 1475696, and could you lend me your system to have try.

**Comment by xiywang** [comment]
@quanwenli Environment provided in the bugzilla.

**Comment by xiywang** [comment]
(1/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.my.openflow_acl_test.access_deny_all_target.ftp: PASS (178.80 s)
  (2/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.my.openflow_acl_test.access_deny_special_target.ftp: PASS (329.60 s)
  (3/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.my.openflow_acl_test.accessed_deny_all_target.ftp: PASS (345.33 s)
  (4/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.my.openflow_acl_test.accessed_deny_special_target.ftp: PASS (328.00 s)

Test passed.

**Comment by xiywang** [comment]
Maybe need to open a new bug to track the issue mentioned in the bugzilla of atbr0 and install dnsmasq on host.

**Comment by xiywang** [comment]
Acked-by: Xiyue Wang xiywang@redhat.com

File: qemu/tests/openflow_acl_test.py [hunk 1]
```diff
@@ -276,9 +277,14 @@ def setup_service(setup_target):
         error_context.context("Set up %s service in %s"
                               % (setup_params.get("service"), setup_target),
                               logging.info)
+        setup_func(setup_cmd, timeout=setup_timeout)
+        if params.get("copy_ftp_site") and setup_target != "localhost":
+            ftp_site = os.path.join(data_dir.get_deps_dir(), params.get("copy_ftp_site"))
+            ftp_dir = params.get("ftp_dir")
+            setup_vm.copy_files_to(ftp_site, ftp_dir)
+
```
Comment (vivianQizhu) [inline qemu/tests/openflow_acl_test.py]: @xiywang Just double confirm, is this what we expect to exchange the order of `setup_cmd` and `prepare_cmd`?

Comment (xiywang) [reply qemu/tests/openflow_acl_test.py]: @vivianQizhu Yes, that's OK.

**Pull request merged**